greenheart/simulation/greenheart_simulation.py

```diff
@@ -33,6 +40,68 @@
 
 pd.options.mode.chained_assignment = None  # default='warn'
 
+# """
+# Recursively converts complex types to JSON/YAML-compatible formats.
+# Handles:
+# - `np.ndarray` -> list
+# - `tuple` -> list
+# - `np.generic` (e.g., `np.float64`, `np.int32`) -> native Python types
+# - `pandas.DataFrame` -> list of dicts, recursively processed
+# - `pandas.Series` -> list, recursively processed
+# - `attrs` objects -> dict of serialized attributes
+# - Handles deeply nested structures
+
+# Note: this function was originally created by ChatGPT and edited manually to work as desired
+# """
+
+
+def convert_to_serializable(value: Any) -> float | int | str | type(None) | list | dict:
+    """Recursively converts complex types to JSON/YAML-compatible formats.
+
+    Handles:
+    - `np.ndarray` -> list
+    - `tuple` -> list
+    - `np.generic` (e.g., `np.float64`, `np.int32`) -> corresponding native Python types
+    - `pandas.DataFrame` -> list of dicts, recursively processed
+    - `pandas.Series` -> list, recursively processed
+    - `attrs` objects -> dict of serialized attributes
+    - Handles deeply nested structures
+
+    Note: this function was originally created by ChatGPT and edited manually to work as desired
+
+    Args:
+        value (Any): value to converted for output to yaml
+    Returns:
+        Union[float, int, str, type(None), list, dict]: input value in yaml-compatible format
+    """
+
+    if isinstance(value, np.generic):
+        # Handles NumPy scalar types, converting to python native types
+        return value.item()
+    elif isinstance(value, (np.ndarray, tuple, list, pd.Series)):
```

Comment:
I missed this before, but these can all be standalone 'if's because each step in the control flow has a `return`. This means the `else` line can be removed with the `return str(value)` being dedented, ensuring the function has a return at the outermost level.

Reply:
Good point, done.